src/hash.h

```diff
@@ -10,32 +10,28 @@
 #include "common.h"
 
 #include "git2/oid.h"
+#include "hash/sha1.h"
 
-typedef struct git_hash_ctx git_hash_ctx;
+typedef enum git_hash_algorithm {
+	GIT_HASH_ALGORITHM_SHA1
+} git_hash_algorithm;
 
-int git_hash_global_init(void);
-int git_hash_ctx_init(git_hash_ctx *ctx);
-void git_hash_ctx_cleanup(git_hash_ctx *ctx);
-
-#if defined(GIT_SHA1_COLLISIONDETECT)
-# include "hash/sha1/collisiondetect.h"
-#elif defined(GIT_SHA1_COMMON_CRYPTO)
-# include "hash/sha1/common_crypto.h"
-#elif defined(GIT_SHA1_OPENSSL)
-# include "hash/sha1/openssl.h"
-#elif defined(GIT_SHA1_WIN32)
-# include "hash/sha1/win32.h"
-#elif defined(GIT_SHA1_MBEDTLS)
-# include "hash/sha1/mbedtls.h"
-#else
-# include "hash/sha1/generic.h"
-#endif
+typedef struct git_hash_ctx {
+	union {
+		git_hash_sha1_ctx sha1;
+	} ctx;
+	git_hash_algorithm alg;
```

Comment:
Maybe put that first ? I'm thinking we might want a `GIT_HASH_ALGORITHM_CUSTOM`, and having that appear first would allow a user-provided struct to be overlayed transparently ?